### ISSUE TITLE
[Performance Issue]: PTXAS issue with Cuda, and out of memory

### ISSUE THREAD

**Issue by roushrsh** [SciSharp/TensorFlow.NET#1019]
### Brief Description

Hi, 

I'm getting the error : 
2023-04-07 15:38:25.431418: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8100
2023-04-07 15:38:25.875660: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] INTERNAL: ptxas exited with non-zero error code -1, output:
Relying on driver to perform ptx compilation.
Modify $PATH to customize ptxas location.
This message will be only logged once.
2023-04-07 15:38:26.216472: I tensorflow/stream_executor/cuda/cuda_blas.cc:1614] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once. 
2023-04-07 15:38:25.431418: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8100
2023-04-07 15:38:25.875660: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] INTERNAL: ptxas exited with non-zero error code -1, output:
Relying on driver to perform ptx compilation.
Modify $PATH to customize ptxas location.
This message will be only logged once.
2023-04-07 15:38:26.216472: I tensorflow/stream_executor/cuda/cuda_blas.cc:1614] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
2023-04-07 15:38:26.260973: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1616] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 21348 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 4090, pci bus id: 0000:01:00.0, compute capability: 8.9
Elapsed Time is 368 ms

-------------------
The same prediction takes ~30 ms on python using batch_predict (10% more with just predict). (b~15000 with cpu on c#, so gpu is definitely being used). 

Possible issue with is ptxas not being recognized (it is in my path, python has no problem finding it):

Also, if I run:

var T3 = model.predict((xx1, xx5), batch_size: 128);
T3 = model.predict((xx1, xx5), batch_size: 128); 
T3 = model.predict((xx1, xx5), batch_size: 128);#out of memory error here  'OOM when allocating tensor with shape..etc'

it runs out of memory by the second batch prediction. I have to reduce it down to 32 for it to make it through all three. Is there a way to empty the memory after each prediction? Or something I must do?

Any suggestions?

### Device and Context

13900k, 4090 RTX

### Benchmark

Python is 5-10x faster

**Comment by roushrsh** [comment]
Batch Predict with Keras.Net is 50ms, still much slower, but ok.Problem there is, everything has to be from NDarrays, which take 250ms to prepare anyways...

**Comment by AsakusaRinne** [comment]
> 

Hi, what is the original data format of you app? The reason that it takes 250ms may be memory copy when constructing NDArray.

**Comment by roushrsh** [comment]
Yes, it is large, 4319*4*1, so roughly equivalent 70*70*3 . Is there any way to bypass this? Prediction only takes 0.3ms on python, 4 orders of magnitude faster. CPU runs twice as fast as GPU because of this.

**Comment by Oceania2018** [comment]
> Batch Predict with Keras.Net is 50ms, still much slower, but ok.Problem there is, everything has to be from NDarrays, which take 250ms to prepare anyways...

Are you using TensorFlow.NET/ TensorFlow.Keras or Keras.NET, they're different library,

**Comment by roushrsh** [comment]
Hi @Oceania2018 , 

I'm using Tensorflow.net
My only imports are:

          using Tensorflow;
          using Tensorflow.Keras.Optimizers;
          using static Tensorflow.Binding;
          using static Tensorflow.KerasApi;
          using static Tensorflow.TensorShapeProto.Types;
          using Tensorflow.NumPy;
          using System.Diagnostics;
          using System.IO;
          using System.Linq;



I have also tried Keras.Net before. it predicts faster when I use their function 'PredictOnBatch' (50ms), however the part where I have to convert to their NDarray format takes 250ms for my 4319*4 image, so it's useless as well.

**Comment by roushrsh** [comment]
Thanks do you have a link to where I can read how to do that, or convert my model? I did try ONNX/MicrosoftML as well and had the same issue.

My model is just a yolo model, 7 inputs:
[4319x1, 4319x1, 4319x1, 4319x1, 1,1,1,1,1]
YOLO CNNs (800,000)
output is a sigmoid regression

**Comment by Oceania2018** [comment]
@roushrsh There is a blog about how to user [AutoGraph](https://medium.com/scisharp/tensorflow-autograph-for-net-a89f1e97a4a5) annotation. You can utilize this approache to optimize the prediction part.

**Comment by AsakusaRinne** [comment]
Hi, it came to my brain today that since tensorflow 2.11, the GPU support for native-windows was dropped. Unfortunately the latest version of tf.net uses tf2.11 library. So if you use latest tf.net on Windows, the performance may be worse than expected.

**Comment by roushrsh** [comment]
Oh wow, that could be it. Should I just test it on linux or similar, or is there a solution?

**Comment by AsakusaRinne** [comment]
> Oh wow, that could be it. Should I just test it on linux or similar, or is there a solution?

Sorry I'm wrong, the tensorflow.redist has CPU version updated to 2.11 but GPU version still 2.10. Please just ignore what I said (still, updating GPU redist to 2.11 would cause that problem but it's not the point of this issue). Would you like to provide an example to reproduce the problem? The performance issue has high priority to us and I'll try to find a solution for it.

**Comment by abrguyt** [comment]
I'm also experiencing this problem - using the Tensorflow.NET / Tensorflow.Keras combo - every prediction seems to leak memory. Even when the NDArray or Tensor objects are explicitly disposed after every prediction.

As this library is not maintained any longer it seems for any long term usage with lots of data this is unworkable.
